src/query/storage/m3/dynamic_cluster.go

```diff
@@ -88,10 +96,397 @@ func NewDynamicClusterOptions() DynamicClusterOptions {
+				NamespaceID: md.ID(),
+				Session:     clusterCfg.session,
+				Retention:   retOpts.RetentionPeriod(),
+			})
+			if err != nil {
+				return nil, err
+			}
+		}
+		clusterNamespaces = append(clusterNamespaces, clusterNamespace)
+	}
+
+	return clusterNamespaces, nil
+}
+
+func (d *dynamicCluster) updateClusterNamespacesWithLock() {
+	nsCount := 0
+	for _, nsMap := range d.namespacesByEtcdCluster {
+		for _, clusterNamespaces := range nsMap.metadataToClusterNamespaces {
+			nsCount += len(clusterNamespaces)
+		}
+	}
+
+	var (
+		newNamespaces            = make(ClusterNamespaces, 0, nsCount)
+		newAggregatedNamespaces  = make(map[RetentionResolution]ClusterNamespace)
+		newUnaggregatedNamespace ClusterNamespace
+	)
+
+	for _, nsMap := range d.namespacesByEtcdCluster {
+		for _, clusterNamespaces := range nsMap.metadataToClusterNamespaces {
+			for _, clusterNamespace := range clusterNamespaces {
+				attrs := clusterNamespace.Options().Attributes()
+				if attrs.MetricsType == storagemetadata.UnaggregatedMetricsType {
+					if newUnaggregatedNamespace != nil {
+						d.logger.Warn("more than one unaggregated namespace found. using most recently "+
+							"discovered unaggregated namespace",
+							zap.String("existing", newUnaggregatedNamespace.NamespaceID().String()),
+							zap.String("new", clusterNamespace.NamespaceID().String()))
+					}
+					newUnaggregatedNamespace = clusterNamespace
+				} else {
+					retRes := RetentionResolution{
+						Retention:  attrs.Retention,
+						Resolution: attrs.Resolution,
+					}
+					existing, ok := newAggregatedNamespaces[retRes]
+					if ok {
+						d.logger.Warn("more than one aggregated namespace found for retention and resolution. "+
+							"using most recently discovered aggregated namespace",
+							zap.String("retention", retRes.Retention.String()),
+							zap.String("resolution", retRes.Resolution.String()),
+							zap.String("existing", existing.NamespaceID().String()),
+							zap.String("new", clusterNamespace.NamespaceID().String()))
+					}
+					newAggregatedNamespaces[retRes] = clusterNamespace
+				}
+			}
+		}
+	}
+
+	newNamespaces = append(newNamespaces, newUnaggregatedNamespace)
+	for _, ns := range newAggregatedNamespaces {
+		newNamespaces = append(newNamespaces, ns)
+	}
+
+	d.unaggregatedNamespace = newUnaggregatedNamespace
+	d.aggregatedNamespaces = newAggregatedNamespaces
+	d.namespaces = newNamespaces
+}
+
+func (d *dynamicCluster) Close() error {
+	d.Lock()
+	defer d.Unlock()
+
+	if d.closed {
+		return errNsWatchAlreadyClosed
+	}
+
+	d.closed = true
+
+	var multiErr xerrors.MultiError
+	for _, watch := range d.nsWatches {
+		if err := watch.Close(); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	}
+
+	if !multiErr.Empty() {
+		return multiErr.FinalError()
+	}
+
+	return nil
+}
+
+func (d *dynamicCluster) ClusterNamespaces() ClusterNamespaces {
+	d.RLock()
+	defer d.RUnlock()
+
+	return d.namespaces
```

Comment:
nit: these don't need to be deferred if you use a var

Reply:
So you mean something like:

```
d.RLock()
namespaces := d.namespaces
d.RUnlock()

return namespaces
```
ya? Also, do we generally prefer to not `defer Unlock()` vs manually doing so once done?

Reply:
Yeah it's probably not great practice (and as of [go 1.14 ](https://golang.org/doc/go1.14)this probably upgrades to _actively bad_ practice), and honestly wouldn't really make a difference outside of very hot paths; maybe we should start looking to kicking this habit and just deferring them more